### PR TITLE
feat(appmesh): add Route matching on path, query parameters, metadata, and method name

### PR DESCRIPTION
**Pull request by Seiya6329**
Adding new match properties for `Route`.
- For HTTP match, adding `path` and `queryParameters`. Remove `prefixPath`.
- For gRPC match, adding `metadata` and `method name`

BREAKING CHANGE: `prefixPath` property in `HttpRouteMatch` has been renamed to `path`, and its type changed from `string` to `HttpRoutePathMatch`

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-appmesh/lib/private/utils.ts [hunk 1]
```diff
@@ -95,13 +97,42 @@ export function renderMeshOwner(resourceAccount: string, meshAccount: string) :
...
+  if (metadata && (metadata.length < MIN_LENGTH || metadata.length > MAX_LENGTH)) {
+    throw new Error(`Number of metadata provided for matching must be between ${MIN_LENGTH} and ${MAX_LENGTH}, got: ${metadata.length}`);
+  }
+}
+
+/**
+ * This is the helper method to validate at least one of gRPC match type is defined.
+ */
+export function validateGprcMatch(match: GrpcRouteMatch) {
+  for (const property in match) {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:131]: Unfortunately, I don't think this does what you want it to do.

TypeScript uses a structural type system, so for example, you could do:

```ts
const myMatch = {
  randomProperty: true,
};
```

and then pass `myMatch` to where `GrpcRouteMatch` is expected, and your method would return `true`.

I think you might have to spell out the condition explicitly, something like `return match.serviceName == undefined && match.methodName === undefined && match.metadata === undefined`.

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts [hunk 2]
```diff
@@ -415,23 +443,33 @@ class HttpRouteSpec extends RouteSpec {
...
     const headers = this.match?.headers;
-    validateMatchArrayLength(headers);
+    const queryParameters = this.match?.queryParameters;
+
+    if (prefixPathMatch && prefixPathMatch[0] !== '/') {
+      throw new Error(`Prefix Path for the match must start with \'/\', got: ${prefixPathMatch}`);
+    }
+    if (wholePathMatch?.exact && wholePathMatch.exact[0] !== '/') {
+      throw new Error(`Exact Path for the match must start with \'/\', got: ${wholePathMatch.exact}`);
+    }
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:459]: Can we actually move this validation to the `HttpRoutePatchMatch` subclasses instead? I think this will greatly simplify this class.

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts [hunk 3]
```diff
@@ -415,23 +443,33 @@ class HttpRouteSpec extends RouteSpec {
...
       },
       match: {
-        prefix: prefixPath,
+        prefix: prefixPathMatch,
+        path: wholePathMatch,
         headers: headers?.map(header => header.bind(scope).headerMatch),
         method: this.match?.method,
         scheme: this.match?.protocol,
+        queryParameters: queryParameters?.map(queryParameter => queryParameter.bind(scope).queryParameterMatch),
       },
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:473]: Here's an idea: let's leave `match` as `undefined` if either `this.match` is `undefined`, or if all of its properties are `undefined`.

So, this would be:

```suggestion
      match: allPropertiesAreUndefined(this.match) ? undefined : {
        prefix: pathMatchConfig.prefixPathMatch,
        path: pathMatchConfig.wholePathMatch,
        headers: headers?.map(header => header.bind(scope).headerMatch),
        method: this.match?.method,
        scheme: this.match?.protocol,
        queryParameters: queryParameters?.map(queryParameter => queryParameter.bind(scope).queryParameterMatch),
      },
```

File: packages/@aws-cdk/aws-appmesh/lib/private/utils.ts [hunk 4]
```diff
@@ -95,13 +97,42 @@ export function renderMeshOwner(resourceAccount: string, meshAccount: string) :
 }
 
 /**
- * This is the helper method to validate the length of match array when it is specified.
+ * This is the helper method to validate the length of HTTP match array when it is specified.
  */
-export function validateMatchArrayLength(headers?: HeaderMatch[]) {
+export function validateHttpMatchArrayLength(headers?: HeaderMatch[], queryParameter?: QueryParameterMatch[]) {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:102]:
```suggestion
export function validateHttpMatchArrayLength(headers?: HeaderMatch[], queryParameters?: QueryParameterMatch[]) {
```

**Review by skinny85 — DISMISSED** [review]
In general, looks great! A few suggestion that I believe will greatly simplify this code.

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts (hunk 3 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:473]: Actually [`match`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-route-httproute.html) is the required property so would not be able to set `undefined`.

In addition, we set `prefix` to `/` if neither `prefix` and `path` are defined so we are guaranteed to have at least one property be defined.

Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:473]: I see all of the properties of the `CfnGatewayRoute.HttpGatewayRouteMatchProperty` are opitonal.

Is there some CloudFormation validation that errors out if an empty object is passed as its value, or is it legal to just omit all of the properties? Because if it's legal, let's just do that.

Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:473]: > Is there some CloudFormation validation that errors out if an empty object is passed as its value

Yes, CloudFormation has a validation to prohibit passing empty value or empty array.

> is it legal to just omit all of the properties

No, CloudFormation also validates if least one type of match is specified.

Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:473]: Since CDK has a validation to check empty array and non-array property accepts an object so I believe CDK-side has all validation that are needed.

File: packages/@aws-cdk/aws-appmesh/README.md [hunk 5]
```diff
@@ -406,6 +406,18 @@ A `route` is associated with a virtual router, and it's used to match requests f
 
 If your `route` matches a request, you can distribute traffic to one or more target virtual nodes with relative weighting.
 
+The _RouteSpec_ class provides an easy interface for defining new protocol specific route specs.
+The `tcp()`, `http()`, `http2()`, and `grpc()` methods provide the spec necessary to define a protocol specific spec.
+
+For HTTP based routes, the match field can be used to match on 
+path (prefix, exact, or regex), HTTP method, scheme, HTTP headers, and query parameters.
+By default, an HTTP based route will match on all routes.
```
[the remark this answers is not in the document]
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/README.md:414]: I agree, request fits better here.

File: packages/@aws-cdk/aws-appmesh/README.md [hunk 6]
```diff
@@ -406,6 +406,18 @@ A `route` is associated with a virtual router, and it's used to match requests f
...
+
+For gRPC based routes, the match field can be used to match on service name, method name, and metadata.
+When specifying the method name, service name must also be specified.
+
+Add an HTTP route that matches based on prefix path.
```
[the remark this answers is not in the document]
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/README.md:419]: Addressed

File: packages/@aws-cdk/aws-appmesh/README.md [hunk 7]
```diff
@@ -511,12 +528,28 @@ router.addRoute('route-grpc-retry', {
...
+      // When method name is specified, service name must be also specified.
+      methodName: 'methodname',
+      serviceName: 'servicename',
+      metadata: [
+        // All specified metadata must match for the route to match.
+        appmesh.HeaderMatch.valueStartsWith('Content-Type', 'application/'),
+        appmesh.HeaderMatch.valueDoesNotStartWith('Content-Type', 'text/'),
+      ],
+    },
+
```
[the remark this answers is not in the document]
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/README.md:547]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  /**
+   * The value of the path must match the specified prefix.
+   *
+   * @param prefix This parameter must always start with /, which by itself matches all requests to the virtual service name.
+   *  You can also match for path-based routing of requests. For example, if your virtual service name is my-service.local
+   *  and you want the route to match requests to my-service.local/metrics, your prefix should be /metrics.
+   */
+  public static startingWith(prefix: string): HttpRoutePathMatch {
```
[the remark this answers is not in the document]
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:55]: Good point. Will change it.

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    return new HttpRoutePathPrefixMatch(prefix);
+  }
+
+  /**
+   * Returns the route path match configuration.
+   */
+  public abstract bind(scope: Construct): HttpRoutePathMatchConfig;
+}
+
+class HttpRoutePathPrefixMatch extends HttpRoutePathMatch {
```
[the remark this answers is not in the document]
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:65]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/private/utils.ts [hunk 10]
```diff
@@ -95,13 +97,42 @@ export function renderMeshOwner(resourceAccount: string, meshAccount: string) :
...
+ */
+export function validateGprcMatch(match: GrpcRouteMatch) {
+  for (const property in match) {
+    if (match.hasOwnProperty(property)) {
+      return;
+    }
+  }
+
+  throw new Error('At least one gRPC match property must be provided');
 }
```
[the remark this answers is not in the document]
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:138]: Well noted.

File: packages/@aws-cdk/aws-appmesh/lib/private/utils.ts (hunk 4 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:102]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts (hunk 2 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:459]: Moved.

File: packages/@aws-cdk/aws-appmesh/lib/private/utils.ts (hunk 1 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:131]: I am not sure if I am understanding your concern here.

I tried above example and when `myMatch` is inserted, I am seeing the compile-time error, "type '{ randomProperty: boolean; }' has no properties in common with type 'GrpcRouteMatch'."

Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:131]: I think I understood the issue here. Looks like if one of the property of the `match` matches with `GrpcRouteMatch` then there would no error be shown.

```ts
const myMatch = {
      serviceName: 'test',  // This exists in GrpcRouteMatch
      hostname: 'GatewayRouteHostnameMatch',  // This does NOT exist in GrpcRouteMatch
    };

    validateGrpcMatch(myMatch);
```

However, would this actually a problem? Since the `validateGrpcMatch` function is private and thus as long as the developer uses it with correct context I feel like it is ok.

Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:131]: It doesn't matter that the function is private - it gets called with a value provided by the customer of the AppMesh module.

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 11]
```diff
@@ -0,0 +1,98 @@
+import { CfnRoute } from './appmesh.generated';
+
+// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
+// eslint-disable-next-line no-duplicate-imports, import/order
+import { Construct } from '@aws-cdk/core';
+
+/**
+ * Path and Prefix properties for HTTP route match.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:8]:
```suggestion
 * The type returned from the `bind()` method in {@link HttpRoutePathMatch}.
```

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+ */
+export interface HttpRoutePathMatchConfig {
+  /**
+   * Route CFN configuration for path match.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:12]:
```suggestion
   * Route configuration for matching on the complete URL path of the request.
```

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+   *
+   * @default - no path match.
+   */
+  readonly wholePathMatch?: CfnRoute.HttpPathMatchProperty;
+
+  /**
+   * String defining the prefix match.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:19]:
```suggestion
   * Route configuration for matching on the prefix of the URL path of the request.
```

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+import { Construct } from '@aws-cdk/core';
+
+/**
+ * Path and Prefix properties for HTTP route match.
+ */
+export interface HttpRoutePathMatchConfig {
+  /**
+   * Route CFN configuration for path match.
+   *
+   * @default - no path match.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:14]:
```suggestion
   * @default - no matching will be performed on the complete URL path
```

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+   */
+  readonly wholePathMatch?: CfnRoute.HttpPathMatchProperty;
+
+  /**
+   * String defining the prefix match.
+   *
+   * @default - no prefix match
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:21]:
```suggestion
   * @default - no matching will be performed on the prefix of the URL path
```

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+   */
+  readonly prefixPathMatch?: string;
+}
+
+/**
+ * Defines HTTP route path or prefix request match.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:27]:
```suggestion
 * Defines HTTP route matching based on the URL path of the request.
```

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+ */
+export abstract class HttpRoutePathMatch {
+  /**
+   * The value of the path must match the specified value exactly.
+   *
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:32]:
```suggestion
   * The provided `path` must start with the '/' character.
   * ```

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+   * @param path The exact path to match on
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:33]:
```suggestion
   * @param path the exact path to match on
```

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+   */
+  public static exactly(path: string): HttpRoutePathMatch {
+    return new HttpRouteWholePathMatch({ exact: path });
+  }
+
+  /**
+   * The value of the path must match the specified regex.
+   *
+   * @param regex The regex used to match the path
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:42]:
```suggestion
   * @param regex the regex used to match the path
```

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+  public static regex(regex: string): HttpRoutePathMatch {
+    return new HttpRouteWholePathMatch({ regex: regex });
+  }
+
+  /**
+   * The value of the path must match the specified prefix.
+   *
+   * @param prefix This parameter must always start with /, which by itself matches all requests to the virtual service name.
+   *  You can also match for path-based routing of requests. For example, if your virtual service name is my-service.local
+   *  and you want the route to match requests to my-service.local/metrics, your prefix should be /metrics.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:53]:
```suggestion
   * @param prefix the value to use to match the beginning of the path part of the URL of the request.
   *   It must start with the '/' character. If provided as "/", matches all requests.
   *   For example, if your virtual service name is "my-service.local"
   *   and you want the route to match requests to "my-service.local/metrics", your prefix should be "/metrics".
```

File: packages/@aws-cdk/aws-appmesh/lib/private/utils.ts (hunk 1 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:131]: @skinny85 - Ok. I think I am understanding the the problem here. However, I still think `return match.serviceName == undefined && match.methodName === undefined && match.metadata === undefined` is unnecessary because the criteria for `validateGrpcMatch` is to validate if "at least one match property is defined" and if `myMatch` example does not contain matching property then then there will be a compile-time error.

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+class HttpRoutePrefixPathMatch extends HttpRoutePathMatch {
+  constructor(
+    private readonly prefix: string,
+  ) {
+    super();
+  }
+
+  bind(_scope: Construct): HttpRoutePathMatchConfig {
+    if (this.prefix && this.prefix[0] !== '/') {
+      throw new Error(`Prefix Path for the match must start with \'/\', got: ${this.prefix}`);
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:74]: I would probably move this validation into the constructor of this class. But not a huge deal.

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+
+class HttpRouteWholePathMatch extends HttpRoutePathMatch {
+  constructor(
+    private readonly match: CfnRoute.HttpPathMatchProperty,
+  ) {
+    super();
+  }
+
+  bind(_scope: Construct): HttpRoutePathMatchConfig {
+    if (this.match.exact && this.match.exact[0] !== '/') {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:90]: I would probably move this validation into the constructor of this class. But not a huge deal.

File: packages/@aws-cdk/aws-appmesh/lib/private/utils.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:131]: Still an issue.

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts [hunk 23]
```diff
@@ -415,23 +443,26 @@ class HttpRouteSpec extends RouteSpec {
...
 
   public bind(scope: Construct): RouteSpecConfig {
-    const prefixPath = this.match ? this.match.prefixPath : '/';
-    if (prefixPath[0] != '/') {
-      throw new Error(`Prefix Path must start with \'/\', got: ${prefixPath}`);
-    }
+    const pathMatchConfig = this.match?.path?.bind(scope);
 
+    // Set prefix path match to '/' if none of path matches are defined.
+    const prefixPathMatch = pathMatchConfig ? pathMatchConfig.prefixPathMatch : '/';
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:449]: No need for this complexity here - you can just do:

```suggestion
    const pathMatchConfig = (this.match?.path ?? HttRoutePathMatch.startsWith('/')).bind(scope);
```

And then below:

```ts
        prefix: pathMatchConfig.prefixPathMatch,
        path: pathMatchConfig.wholePathMatch,
```

File: packages/@aws-cdk/aws-appmesh/lib/private/utils.ts (hunk 1 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:131]: To add a note, if we are doing the explicit check, I think it should check whether they are not empty rather than checking if they not `undefined`.

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts [hunk 24]
```diff
@@ -415,23 +443,26 @@ class HttpRouteSpec extends RouteSpec {
...
+    validateHttpMatchArrayLength(headers, queryParameters);
 
     const httpConfig: CfnRoute.HttpRouteProperty = {
       action: {
         weightedTargets: renderWeightedTargets(this.weightedTargets),
       },
       match: {
-        prefix: prefixPath,
+        prefix: prefixPathMatch,
+        path: pathMatchConfig?.wholePathMatch,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:461]: There's one subtlety with the way you've written this method.

If a customer provides a `match` without a `path`, but with a different property, like `method`:

```ts
match: {
  method: HttRouteMethod.GET,
},
...
```

You will render a `prefix` of `'/'` in the template, even though there's no need for it (the service would set that default itself).

Not sure if this is something you want to fix - I'm not intending to block the PR on it. But if you wanted to, you could do it pretty easily with something like this:

```
    const pathMatchConfig = (allMatchPropertiesAreUndefined(this.match)
      ? HttRoutePathMatch.startsWith('/')
      : this.match?.path)?.bind(scope);
```

In other words - only setting the `path` property if no other `match` property was set.

Up to you!

**Review by skinny85 — CHANGES_REQUESTED** [review]
Looks great @Seiya6329! Just a few finishing touches, and this will be ready for merging.

File: packages/@aws-cdk/aws-appmesh/lib/private/utils.ts (hunk 1 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/private/utils.ts:131]: Concluded to perform explicit check to ensure passed object contains necessary properties.

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts (hunk 11 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:8]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts (hunk 13 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:19]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts (hunk 12 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:12]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts (hunk 14 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:14]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts (hunk 15 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:21]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts (hunk 16 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:27]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts (hunk 17 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:32]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts (hunk 18 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:33]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts (hunk 19 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:42]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts (hunk 20 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:53]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts (hunk 23 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:449]: This is great know. Thanks!

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts (hunk 24 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:461]: I think we can keep it this way. 

If we would like be more explicit on assigning `/` to `path` only if none of the match properties are defined, it could be more straightforward by just requiring to specify one of the match property instead of using the default.

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts (hunk 21 above)
Comment (Seiya6329) [reply packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:74]: Addressed

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+  /**
+   * Returns the route path match configuration.
+   */
+  public abstract bind(scope: Construct): HttpRoutePathMatchConfig;
+}
+
+class HttpRoutePrefixPathMatch extends HttpRoutePathMatch {
+  constructor(
+    private readonly prefix: string,
+  ) {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:70]:
```suggestion
  constructor(private readonly prefix: string) {
```

File: packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+    return {
+      prefixPathMatch: this.prefix,
+    };
+  }
+}
+
+class HttpRouteWholePathMatch extends HttpRoutePathMatch {
+  constructor(
+    private readonly match: CfnRoute.HttpPathMatchProperty,
+  ) {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/http-route-path-match.ts:88]:
```suggestion
  constructor(private readonly match: CfnRoute.HttpPathMatchProperty) {
```

File: packages/@aws-cdk/aws-appmesh/lib/route-spec.ts [hunk 27]
```diff
@@ -415,23 +443,25 @@ class HttpRouteSpec extends RouteSpec {
...
+    validateHttpMatchArrayLength(headers, queryParameters);
 
     const httpConfig: CfnRoute.HttpRouteProperty = {
       action: {
         weightedTargets: renderWeightedTargets(this.weightedTargets),
       },
       match: {
-        prefix: prefixPath,
+        prefix: pathMatchConfig.prefixPathMatch,
+        path: pathMatchConfig?.wholePathMatch,
```
Comment (skinny85) [inline packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:460]:
```suggestion
        path: pathMatchConfig.wholePathMatch,
```

**Review by skinny85 — DISMISSED** [review]
Looks great @Seiya6329!

**Review by skinny85 — APPROVED** [review]

**Comment by Seiya6329** [comment]
@skinny85 - Thank you very much for making the change on the last suggestions!
